File: 473 Matchsticks to Square/473 Matchsticks to Square.go
func makesquare(matchsticks []int) bool {
    sum := 0
	for _, n := range matchsticks {
		sum += n
	}
	if sum%4 != 0 || sum == 0 {
		return false
	}
	target := sum / 4
	sort.Slice(matchsticks, func(i, j int) bool { return matchsticks[i] > matchsticks[j] })
	var dfs func(curSum int, side int, i int) bool
	dfs = func(curSum int, side int, idx int) bool {
		if side == 4 {
			return true
		}
		if curSum == target {
			return dfs(0, side+1, 0)
		}
		if curSum > target {
			return false
		}
		for i := idx; i < len(matchsticks); i++ {
			if matchsticks[i] != 0 {
				tmp := matchsticks[i]
				matchsticks[i] = 0
				if dfs(curSum+tmp, side, i+1) {
					return true
				}

				matchsticks[i] = tmp
                if curSum == 0 {
                    return false
                }
			}
		}
		return false
	}
	return dfs(0, 1, 0)
}

File: 473 Matchsticks to Square/473 Matchsticks to Square.py
"""
You are given an integer array matchsticks where matchsticks[i] is the length of the ith matchstick. You want to use all the matchsticks to make one square. You should not break any stick, but you can link them up, and each matchstick must be used exactly one time.
Return true if you can make this square and false otherwise.

Example 1:

Input: matchsticks = [1,1,2,2,2]
Output: true
Explanation: You can form a square with length 2, one side of the square came two sticks with length 1.

Example 2:

Input: matchsticks = [3,3,3,3,4]
Output: false
Explanation: You cannot find a way to form a square with all the matchsticks.
"""
class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        # backtracking
        s = sum(matchsticks)
        n = len(matchsticks)
        if s % 4 != 0:
            return False
        target = s // 4
        matchsticks.sort(reverse=True)
        # @lru_cache(None)
        def dfs(a, b, c, d, i):
            if i == n:

                return True
                # if a == target and b == target and c == target:
                #     return True
                # else:
                #     return False
            if a + matchsticks[i] <= target and dfs(a + matchsticks[i], b, c, d, i + 1):
                    return True
            if b != a and b + matchsticks[i] <= target and dfs(a, matchsticks[i]+ b, c, d, i + 1):
                    return True
            if b != c and c + matchsticks[i] <= target and dfs(a,  b, c + matchsticks[i], d, i + 1):
                return True
            if c != d and d + matchsticks[i] <= target and dfs(a,  b, c, d + matchsticks[i], i + 1):
                return True
            return False

        return dfs(0,0,0,0,0)
